From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Narayan Reddy <narayanr@nvidia.com>
Date: Wed, 13 Apr 2022 07:15:18 +0000
Subject: [PATCH 031/243] NVIDIA: SAUCE: net: phy: aqr: add phy_mode support

provide phy_mode support in PHY DT.
read phy_mode from DT and apply the
configuration accordingly as per the
platform requirement.

if phy_mode not specified AQR operates
in 10G XFI mode

Bug 3599876

Change-Id: Ife3afc5bd5c3a147cce0c3ea6921474db5382f29
Signed-off-by: Narayan Reddy <narayanr@nvidia.com>
Signed-off-by: Revanth Kumar Uppala <ruppala@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2909746
Tested-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>

diff --git a/drivers/net/phy/aquantia/aquantia_main.c b/drivers/net/phy/aquantia/aquantia_main.c
index 3f8d2a614690c..56752e0d277fa 100644
--- a/drivers/net/phy/aquantia/aquantia_main.c
+++ b/drivers/net/phy/aquantia/aquantia_main.c
@@ -142,6 +142,9 @@
 #define AQR107_OP_IN_PROG_SLEEP		1000
 #define AQR107_OP_IN_PROG_TIMEOUT	100000
 
+#define VEND1_GLOBAL_MDIO_PHYXS_PROV2		0xC441
+#define VEND1_GLOBAL_MDIO_PHYXS_PROV2_USX_AN		BIT(3)
+
 #define VEND1_SEC_INGRESS_CNTRL_REG1		0x7001
 
 #define VEND1_GLOBAL_SYS_CONFIG_SGMII		(BIT(0) | BIT(1) | BIT(3))
@@ -236,9 +239,11 @@ static int aqr_set_mdix(struct phy_device *phydev, int mdix)
 
 static int aqr_config_aneg(struct phy_device *phydev)
 {
+	struct device_node *node = phydev->mdio.dev.of_node;
 	bool changed = false;
 	u16 reg;
-	int ret;
+	int ret, err;
+	int phy_mode;
 
 	ret = aqr_set_mdix(phydev, phydev->mdix_ctrl);
 	if (ret < 0)
@@ -286,6 +291,23 @@ static int aqr_config_aneg(struct phy_device *phydev)
 	if (ret > 0)
 		changed = true;
 
+	err = of_property_read_u32(node, "aquantia,phy_mode", &phy_mode);
+	if (!err) {
+		if (phy_mode == 1) {
+			phydev_info(phydev, "Configuring AQR PHY to 5G Mode\n");
+			phy_write_mmd(phydev, MDIO_MMD_VEND1, VEND1_GLOBAL_CFG_2_5G, 0x0106);
+			phy_write_mmd(phydev, MDIO_MMD_VEND1, VEND1_GLOBAL_CFG_5G, 0x0106);
+			phy_write_mmd(phydev, MDIO_MMD_VEND1, VEND1_GLOBAL_CFG_10G, 0x0000);
+			/* Disable 10G advertizement and restart autoneg */
+			phy_write_mmd(phydev, MDIO_MMD_AN, MDIO_AN_10GBT_CTRL, 0x01E1);
+			/* restart auto-negotiation */
+			genphy_c45_restart_aneg(phydev);
+			phy_write_mmd(phydev, MDIO_MMD_PHYXS, VEND1_GLOBAL_MDIO_PHYXS_PROV2, 0x8);
+		}
+	} else {
+		phydev_info(phydev, "No AQR phy_mode setting in DT\n");
+	}
+
 	return genphy_c45_check_and_restart_aneg(phydev, changed);
 }
 
-- 
2.43.0

